src/component/handlers/composite-android/DraftEditorCompositionHandlerAndroid.js

```diff
@@ -0,0 +1,198 @@
+  }
+};
+
+function findCoveringIndex(contentState, selection, textToFind) {
+  if (!selection.isCollapsed()) {
+    // Expected a collapsed selection, return early
+    return selection;
+  }
+
+  const block = contentState.getBlockForKey(selection.getStartKey());
+  const blockText = block.getText();
+  const selStartOffset = selection.getStartOffset();
+  const matchStartOffset = blockText.indexOf(
+    textToFind,
+    // The earliest textToFind could start is its length before the selection
+    selStartOffset - textToFind.length,
+  );
+
+  if (
+    // Ensure the match exists and contains the selection
+    matchStartOffset >= 0 &&
+    matchStartOffset <= selStartOffset &&
+    matchStartOffset + textToFind.length >= selStartOffset
+  ) {
+    return selection.merge({
+      anchorOffset: matchStartOffset,
+      focusOffset: matchStartOffset + textToFind.length,
+    });
+  }
+
+  console.warn(
+    `findCoveringIndex: couldn't find index of '${textToFind}' in '${blockText}' after offset ${selStartOffset}!`,
+  );
+  return selection;
+}
+
+var DraftEditorCompositionHandlerAndroid = {
+  /**
+   * A `compositionstart` event has fired while we're still in composition
+   * mode. Continue the current composition session to prevent a re-render.
+   */
+  onCompositionStart: function(editor: DraftEditor, e: SyntheticCompositionEvent): void {
+    resetCompositionData();
+    const editorNode = ReactDOM.findDOMNode(editor.refs.editorContainer);
+
+    mutationObserver.observe(editorNode, {childList: true, subtree: true});
+    compositionText = e.data;
+    compositionRange = getCompositionRange(editor, compositionText);
+  },
+
+  onCompositionUpdate: function(editor: DraftEditor, e: SyntheticCompositionEvent): void {
+    if (!hasInsertedCompositionText) {
+      compositionText = e.data;
+      compositionRange = getCompositionRange(editor, compositionText);
+    }
+  },
+
+  onCompositionEnd: function(editor: DraftEditor, e: SyntheticCompositionEvent): void {
+    if (!hasMutation) {
+      handleMutations(mutationObserver.takeRecords());
+    }
+
+    const newText = e.data;
+    if (newText === compositionText) {
+      const nextEditorState = EditorState.acceptSelection(editor._latestEditorState, compositionRange);
+      editor.setMode('edit');
+      editor.update(
+            EditorState.set(nextEditorState, {inCompositionMode: false}),
+        );
+    } else {
+      // If any children have been added/removed the reconciler will crash unless we restore the dom.
+      const mustReset = hasMutation;
+      if (mustReset) {
+        editor.restoreEditorDOM();
+      }
+
+      const nextEditorState = replaceText(
+        editor._latestEditorState,
+        newText,
+        compositionRange,
+        editor._latestEditorState.getCurrentInlineStyle(),
+      );
+
+      editor.setMode('edit');
+      const editorStateProps = mustReset ? {
+        // TODO this is done in the draft composition handler, but I am not sure we should.
```

Comment:
I looked into it, I think that we should. The `nativelyRenderedContent` flag is supposedly to tell draft that the DOM already looks like this editorState and no rerender is necessary. It should always be safe for us to rerender on compositionend though.

Reply:
anything actionable here for the pr ?

Reply:
It's an interesting feature that I didn't realize existed, but I think the only takeaway for this PR is that we can delete this comment.